### PR TITLE
feat: add distribution for the Tuist App

### PR DESCRIPTION
**Pull request by fortmarek**
### Short description 📝

This PR sets up an automatic release and distribution of the Tuist App.

For automatic releases, we're using the [git-cliff](https://git-cliff.org/) tool as we do in other repos, such as XcodeGraph

For distribution, we leverage the [Sparkle](https://sparkle-project.org/) project.

The release roughly goes like this:
- `git-cliff` is run on `main` to determine if a new version of Tuist App should be released (and if so, it determines which version number should be used)
- The version number in the `Project.swift` is updated
- The `mise run bundle:app` script is run -> this script builds, signs, and notarizes the application. The end product is a `.dmg` file.
- `generate_appcast` CLI from Sparkle is run to update the `appcast.xml` file that's expected to be in the `app` directory in git.
- A new release is created with the attached `.dmg` file. The download URL in the `appcast.xml` points to the `.dmg` release artifact.
- Once released, the macOS menu bar app can be updated as it points to the `appcast.xml` in `main`. Sparkle does the rest here.

The only downside of leveraging GitHub releases is that we can't leverage [delta updates](https://sparkle-project.org/documentation/delta-updates/). Since the macOS menu bar app size is relatively small and after a couple of initial weeks, we expect the app to be mostly stable, I don't think missing delta updates are a huge concern.

### How to test the changes locally 🧐

I tested the flow E2E to ensure updates work as expected. It's a tad difficult to test changes locally but you can follow the `Test Sparkle out` section from the [docs](https://sparkle-project.org/documentation/).

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase
- [x] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -29,7 +29,6 @@ var targets: [Target] = [
         name: "TuistCore",
         dependencies: [
             pathDependency,
-            "ProjectDescription",
```
Comment (fortmarek) [inline Package.swift:32]: The Tuist App depends on `TuistCore` and `TuistSupport`. If we kept the `ProjectDescription`, the Tuist App would have to unnecessarily embed the `ProjectDescription`. The usages of `ProjectDescription` in these targets were minimal, so I moved the necessary code to downstream targets.

File: .github/workflows/conventional-pr.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+name: conventional-pr
```
Comment (pepicrft) [inline .github/workflows/conventional-pr.yml:1]: It's ok for now, but I'm a bit afraid about the complexity of maintaining a monorepo setup. For example, we enforce conventions only for app PRs, consider `[Release] ..` commits to refer to app releases but not CLI releases, and mix CLI and App releases.
I'd try to sketch what an iteration of our current architecture would look like to:
1. Support the development of some code in private
2. Support the development of the app in a different repository
3. Eliminate the assumption from the existing code that the consumer is a CLI.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  pull-requests: read
+  statuses: write
+  packages: write
+
+jobs:
+  release:
+    name: Release Tuist App
+    runs-on: macos-latest
+    timeout-minutes: 50
+    if: ${{ !contains(github.event.head_commit.message, '[Release]') }}
```
Comment (pepicrft) [inline .github/workflows/release.yml:24]: Should we make the prefix more specific?

```suggestion
    if: ${{ !contains(github.event.head_commit.message, '[Release] Tuist App') }}
```

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            echo "should-release=true" >> $GITHUB_ENV
+          fi
+      - name: Get next version
+        working-directory: app
+        id: next-version
+        if: env.should-release == 'true'
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          NEXT_VERSION=$(git cliff --include-path "app/**/*" --repository "../" --bumped-version)
```
Comment (pepicrft) [inline .github/workflows/release.yml:63]: Since the `working-directory` here is `app`. Shouldn't `--include-path` be unnecessary?

```suggestion
          NEXT_VERSION=$(git cliff --repository "../" --bumped-version)
```

File: .mise/tasks/install [hunk 5]
```diff
@@ -1,6 +1,8 @@
...
-# mise description="Performs additional tasks that are necessary to work in this repository"
+# mise description="Install all necessary dependencies"
 
 set -euo pipefail
 
-pnpm install -C docs/
\ No newline at end of file
+pnpm install -C $MISE_PROJECT_ROOT/docs/
+swift package --package-path $MISE_PROJECT_ROOT resolve
+swift package --package-path $MISE_PROJECT_ROOT/app resolve
```
Comment (pepicrft) [inline .mise/tasks/install:8]: Out of curiosity, what motivated the creation of another `Project` instead of having a single one?

File: Sources/TuistLoader/Models/PlatformTestInfo.swift [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    /// Global configuration of which platform versions used during mock generation
+    public var PLATFORM_TEST_VERSION: [Platform: String] = [ // swiftlint:disable:this identifier_name
+        .iOS: "11.0",
+        .macOS: "10.15",
+        .watchOS: "8.5",
+        .tvOS: "11.0",
+        .visionOS: "1.0",
+    ]
+
+#endif
```
Comment (pepicrft) [inline Sources/TuistLoader/Models/PlatformTestInfo.swift:14]: What about the following?

```suggestion
#if DEBUG
    import Foundation
    import ProjectDescription
    
    extension Platform {
      func testVersion() -> String {
        switch self {
          case .iOS: return "11.0"
          case .macOS: return "10.15"
          case .watchOS: return "8.5"
          case .tvOS: return "11.0"
          case .visionOS: return "1.0"
        }
      }
    }
#endif
```

File: app/Project.swift [hunk 7]
```diff
@@ -50,6 +54,10 @@ let project = Project(
                     "DEVELOPMENT_TEAM": "U6LC622NKF",
                     "CODE_SIGN_STYLE": "Automatic",
                     "CODE_SIGN_IDENTITY": "Apple Development",
+                ],
+                release: [
+                    "OTHER_CODE_SIGN_FLAGS": "--timestamp --deep",
```
Comment (pepicrft) [inline app/Project.swift:59]: Why are these flags needed? Worth adding a note about them above the line.

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        run: |
+          echo "RELEASE_NOTES<<EOF" >> "$GITHUB_OUTPUT"
+          git cliff --include-path "app/**/*" --repository "../" --unreleased >> "$GITHUB_OUTPUT"
+          echo "EOF" >> "$GITHUB_OUTPUT"
+      - name: Update version
+        working-directory: app
+        if: env.should-release == 'true'
+        run: |
+          sed -i '' -e "s/CFBundleShortVersionString.*/CFBundleShortVersionString\": \"${{ steps.next-version.outputs.NEXT_VERSION_NUMBER }}\",/g" "Project.swift"
+          sed -i '' -e "s/CFBundleVersion.*/CFBundleVersion\": \"${{ steps.next-version.outputs.NEXT_VERSION_NUMBER }}\",/g" "Project.swift"
```
Comment (pepicrft) [inline .github/workflows/release.yml:86]: What do you think if we pass the version information through an environment variable `TUIST_APP_VERSION` that we read from the manifest, and default to `0.1.0` if the variable is absent. That way we don't need this logic, which is a bit brittle.

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        working-directory: app
+        id: next-version
+        if: env.should-release == 'true'
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          NEXT_VERSION=$(git cliff --include-path "app/**/*" --repository "../" --bumped-version)
+
+          if [ "$NEXT_VERSION" = "0.1.0" ]; then
+            NEXT_VERSION="app@0.1.0"
```
Comment (pepicrft) [inline .github/workflows/release.yml:66]: Why is this needed?

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+          echo "NEXT_VERSION_NUMBER=${NEXT_VERSION#*@}" >> "$GITHUB_OUTPUT"
+      - name: Get release notes
+        id: release-notes
+        if: env.should-release == 'true'
+        working-directory: app
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          echo "RELEASE_NOTES<<EOF" >> "$GITHUB_OUTPUT"
+          git cliff --include-path "app/**/*" --repository "../" --unreleased >> "$GITHUB_OUTPUT"
```
Comment (pepicrft) [inline .github/workflows/release.yml:79]: Similar comment here. Do we need `--include-path` considering the working-directory is already `app`?
```suggestion
          git cliff --repository "../" --unreleased >> "$GITHUB_OUTPUT"
```

File: .github/workflows/release.yml [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+          tagging_message: ${{ steps.next-version.outputs.NEXT_VERSION }}
+          commit_message: "[Release] Tuist App ${{ steps.next-version.outputs.NEXT_VERSION }}"
+      - name: Create GitHub Release
+        uses: softprops/action-gh-release@v2
+        if: env.should-release == 'true'
+        with:
+          draft: false
+          repository: tuist/tuist
+          name: ${{ steps.next-version.outputs.NEXT_VERSION }}
+          tag_name: ${{ steps.next-version.outputs.NEXT_VERSION }}
```
Comment (pepicrft) [inline .github/workflows/release.yml:128]: Shouldn't we prefix this one to disambiguate [here](https://github.com/asdf-community/asdf-tuist/blob/a2fabdd3be6fa2ff1a1d5d2a6b9c1a05cb03c491/lib/utils.bash#L20)?

File: .github/workflows/release.yml (hunk 4 above)
Comment (fortmarek) [reply .github/workflows/release.yml:63]: No, at least not per `git-cliff` docs: https://git-cliff.org/docs/usage/monorepos/

File: .mise/tasks/install (hunk 5 above)
Comment (fortmarek) [reply .mise/tasks/install:8]: I see the Tuist App and the Tuist CLI as separate projects. If we do decide to move it away into a different repo, having that distinction early on makes things easier.

File: Sources/TuistLoader/Models/PlatformTestInfo.swift (hunk 6 above)
Comment (fortmarek) [reply Sources/TuistLoader/Models/PlatformTestInfo.swift:14]: Sure, I just copy-pasted what we already had and haven't made any changes.

File: app/Project.swift (hunk 7 above)
Comment (fortmarek) [reply app/Project.swift:59]: These are required for the subsequent notarization.

File: .github/workflows/release.yml (hunk 10 above)
Comment (fortmarek) [reply .github/workflows/release.yml:79]: This is recommended by `git-cliff`, so I wouldn't do this

File: .github/workflows/release.yml (hunk 11 above)
Comment (fortmarek) [reply .github/workflows/release.yml:128]: the `NEXT_VERSION` output already includes a prefix.

File: .github/workflows/release.yml (hunk 9 above)
Comment (fortmarek) [reply .github/workflows/release.yml:66]: `git-cliff` defaults to `0.1.0` if it doesn't find a related tag, but we want the version to be prefixed. See the behavior description [here](https://git-cliff.org/blog/2.0.0#-bump-improvements)

File: .github/workflows/conventional-pr.yml (hunk 2 above)
Comment (fortmarek) [reply .github/workflows/conventional-pr.yml:1]: I'm leaning to a monorepo for the CLI and the app, at least for now, so we don't have to do too many cross-repo changes. If we were to move app into its own repo, we'd need to create an _extra_ repo for the common code across the CLI and the App. This is because App depends on `TuistSupport` and we don't want to get into a situation where we can't merge a PR in the `App` because we need to wait for a new release of `TuistSupport` that's in the CLI that has a weekly release cycle.

Comment (pepicrft) [reply .github/workflows/conventional-pr.yml:1]: We are aligned then. That comment was more toward the future to keep an eye on the complexity that can emerge from this setup.

File: .github/workflows/release.yml (hunk 4 above)
Comment (pepicrft) [reply .github/workflows/release.yml:63]: Understood. I found it a bit misleading.

**Review by pepicrft — APPROVED** [review]
